src/lsm/k_way_merge.zig

```diff
@@ -24,207 +29,260 @@ pub fn KWayMergeIteratorType(
 
-        context: *Context,
-        streams_count: u32,
-        direction: Direction,
-        state: enum { loading, iterating },
+    return struct {
 
-        /// Array of keys, with each key representing the next key in each stream.
-        ///
-        /// `keys` is *almost* structured as a binary heap — to become a heap, streams[0] must be
-        /// peeked and sifted (see pop_internal()).
-        ///
-        /// * When `direction=ascending`, keys are ordered low-to-high.
-        /// * When `direction=descending`, keys are ordered high-to-low.
-        /// * Equivalent keys are ordered from high precedence to low.
-        keys: [streams_max]Key = undefined,
+        // K-way merge via a loser tree algorithm (Knuth Volume 3 p. 253).
+        // Merges k sorted streams using a tournament (loser) tree.
+        // The current global winner lives in `contender`. Internal nodes in `losers`
+        // store the losers of the last comparisons along the root-to-leaf paths.
+        //     0 (winner, contender)
+        //
+        //     1
+        //    / \
+        //   2   3
+        //  / \ / \
+        // 4  5 6  7
+        // -------------
+        // K input streams
+        //
+        // The internal nodes are organized in a flat Eytzinger layout.
+        // That is the tree above is stored as [1][2][3][4][5][6][7].
+        //
+        // There are also a few optimizations that seem to be helpful, but did not work, such as:
+        // - Only store stream_id in the inner nodes, and have a heads array for the first keys
+        //   of the streams. This reduces space (densely packed) and the bytes in swap.
+        //   Unfortunately, it made the code much slower.
 
-        /// For each key in keys above, the corresponding index of the stream containing that key.
-        /// This decouples the order and storage of streams, the user being responsible for storage.
-        /// The user's streams array is never reordered while keys are swapped, only this mapping.
-        streams: [streams_max]u32 = undefined,
+        const KWayMergeIterator = @This();
 
-        /// The number of streams remaining in the iterator.
-        k: u32 = 0,
+        const node_max: u32 = std.math.ceilPowerOfTwoAssert(u32, options.streams_max);
+        const stream_id_invalid = std.math.maxInt(@TypeOf(options.streams_max));
+        const sentinel: Node = .{
+            .key = std.math.maxInt(Key),
+            .stream_id = stream_id_invalid,
+            .sentinel = true,
+        };
 
-        key_popped: ?Key = null,
+        state: enum { loading, iterating },
+        tree_height: u16,
+        nodes_count: u16,
+        streams_count: u16,
+        streams_active: u16,
+        context: *Context,
+        contender: Node,
+        losers: [node_max]Node,
+        direction: Direction,
+        key_popped: ?Key,
+
+        const Node = struct {
+            key: Key,
+            stream_id: u32,
+            sentinel: bool,
```

Comment:
Is there much gain in having a `sentinel: bool` field instead of an `inline fn sentinel(...) bool`, like we do for composite keys?